### ISSUE TITLE
FR: Metadata Hash-ID Storage Option

### ISSUE THREAD

**Issue by ryanwwest** [koreader/koreader#10892]
Currently, highlights, reading progress and the like are stored in Sidecar directories whose location is controlled by setting `Cog -> Document -> Book metadata location` and there are two storage choices:
1. book folder (creates a `book-filename.sdr/` folder alongside the document file)
2. koreader/docsettings/ (within this folder, creates full absolute path folder hierarchies to every book with metadata (so I believe every book ever opened by KOReader)

(2) is nice since it declutters your directories, but as its [parent issue](https://github.com/koreader/koreader/issues/9265) points out, has the negative of losing book progress/highlights if a user renames the book filename or any of its parent directories, or moves the book. This makes this option significantly less useable since it is so inflexible. (2) importantly relies on the absolute path and filename to identify a book.

I'd like to propose a third, non-default option for metadata storage, added to "Book metdata location" above: flat hash-based sidecar folder storage within `koreader/docsettings/`. While the _root location_ doesn't change, the identification method to link a new document does. Rather than using the absolute filepath+filename, this option would use two 'hashing' methods:

- A. For most filetypes, something like an MD5 hash could suffice. This is void if file contents are likely to change, but I think they rarely are (and if a user has a unique different scenario, they can just not pick this third metadata storage option)
- B. For PDFs, there is semi-standard support to save annotations into the PDF files, which would change the MD5. However, PDF files contain a [digital fingerprint](https://web.hypothes.is/help/how-to-save-copy-of-pdf-with-a-different-fingerprint/) that doesn't change from these annotations, so this could instead be used. This isn't really a 'hash' but provides a similar benefit. Hypothes.is actually uses PDF fingerprints for the same purpose as I'm proposing: to decouple the document metadata from a particular 'instance'/'copy' of that document in a particular file.

Both of these IDs are easy to calculate but would not change no matter the location nor name of your book files. This would not only improve on option (1) above by decluttering directories, but also allow the user to freely organize and rename their files as they wish and not risk forgetting to move the `.sdr` folder too. The book metadata is effectively decoupled from a particular file containing that book. This has the additional benefit of letting users sync the `koreader/docsettings/` folder across devices and open any file accessed on one device on any other device, as long as the MD5/fingerprints match.

I recognize this is not as simple as just moving the location of the `.sdr` folders. I would imagine they could all be flatly stored as folders in `koreader/docsettings/` but the folder names would need to change (do you use the hash value or the original book filename? But then how do you disambiguate different files with the same filename from different directories)? I feel like a solution is within reach, though.

(Status update edits: @ryanwwest looked into using PDF IDs for a better ID below but couldn't figure out how to get them into muPDF. He also learned that epub files support a similar identifier. But for now, hash-based storage is only based on the partial MD5 function)

**Comment by NiLuJe** [comment]
Usual caveat that calibre *will* modify files in flight during Send To Device, so multi-device workflows with calibre will almost never have matching content hashes.

**Comment by ryanwwest** [comment]
> Usual caveat that calibre _will_ modify files in flight during Send To Device, so multi-device workflows with calibre will almost never have matching content hashes.

I see. Do you know why they're modified? From [this website](url) I see the following 'update metadata in saved copies' example (which looks like it can be disabled), which might not be the only reason:

![image](https://github.com/koreader/koreader/assets/31378238/e02759a7-ef7d-4636-a47a-6935597b2e29)

So either MD5 hashes are incompatible with Calibre's Send to Device or Calibre needs to be configured before it can be used with this third metadata storage option.

**Comment by NiLuJe** [comment]
Among various other things, by default, calibre doesn't bake any of the metadata changes made in its UI into the books stored in its database, that only makjes it during the send. There are various ways to tweak that, but that's generally extremely non-default and tricky setups.

**Comment by ryanwwest** [comment]
I discovered a different issue while inspecting `book.sdr/metadata.fileformat.lua`: the `doc_path` field contains the absolute path to the associated file, so that would probably best be updated if metadata were stored in this way. In practice, this path hasn't seemed to matter as I use Synthing to sync all of my docs between ~6 devices running KOReader each with a different absolute path, and had never noticed an issue other than occasional syncthing conflicts. Maybe it's because it seems when the sidecar file is loaded, it is getting the document path from the file browser and [it just overwrites whatever was stored in the metdata upon load](https://github.com/koreader/koreader/blob/f78a8c5315d531b6106e878774cf27cc26e1eaa5/frontend/docsettings.lua#L223C19-L223C19), but there are probably more things happening.

Maybe I will try diving into the code to see how to implement something like this third option.

**Comment by Galunid** [comment]
@ryanwwest https://github.com/koreader/koreader/issues/6491#issuecomment-769793624 may be of interest. I'm not sure if it solves the different hashes issue 100% of  the time, or if hash changes after Calibre restart.

**Comment by hius07** [comment]
> the `doc_path` field contains the absolute path to the associated file

Currently it is not used anywhere in KORreader but we keep it updated in all file/book operations.
It is used by KoHighlights.

**Comment by ryanwwest** [comment]
@Galunid Thanks, the discussion you linked seems to point to the 'update metadata in saved copies' checkbox in the screenshot above or similarly named settings as the, or a, culprit for MD5 changes. Either way, I think the initial version of this would just warn users to not use this metadata storage option if using Calibre Send to Device. I'd be curious how many people use that Calibre option.

@hius07 Good to know—maybe it would be good to keep then, though it doesn't mean much in my idealized world of having the same files synced across many devices with different absolute paths. But most probably aren't doing the same thing.

**Comment by poire-z** [comment]
I see from #10901 that you're going at it.
So, a few points:

----
We currently use some hash in the statistics plugin (which don't depend on the filepath, but on the book title/authors/md5 - so as title/authors would change the md5, on the md5 :).
But we don't compute a real md5 sum of the whole book. We do something that I assume is quicker:

https://github.com/koreader/koreader/blob/1f03e76aea5d444a731672e7a593416599d71efb/plugins/statistics.koplugin/main.lua#L429-L450

The KOSync plugin does that too.
But we do it once, and store it in DocSettings.

I also don't know how this "partial md5" plays with PDF when we embed more and more highlights. It may change as we embed more and more highlights. I guess it may not affect us because we compute it at first opening from the start of the PDF at that time - we don't need to recompute later and the PDF content/md5 can change. (But I guess if you reset docsettings, you're now with a new "partial md5" and your reading statistics are lost.)

----
If you want to use such a md5 to name and locate the docsettings, there will be some things that will get slower:
- when you browse in FileBrowser, we need to locate if a docsetting exists (to at least show the little dogear that the book has been opened - or to show other things like reading status - or to check if a custom cover file has been set) : for this, you will have to compute the md5 of all the books you meet while browsing to get to know if they have a docsetting :/
- also, some filebrowser sort options (and probably other things) need to know the above for _all_ books in a directory before it can order them and show the first page - so the need to compute the md5 of all the book :/

----
Dunno how happy @hius07 may be about having a third option for docsettings (we have "doc" and "dir") with all the current if/else - but he may say, given the brain and work he has put in that the last months, if the code has been fully updated to have these checks all internal in DocSettings and the other code does not have to know, so that if you add that 3rd option, it's going to be only in docsettings.lua - and you would not need to touch anything else.

**Comment by ryanwwest** [comment]
@poire-z partialMd5 looks like a good faster replacement for a full MD5 hash (samples 10kb of I think any sized file to hash) and I don't think that particular function also relies on book title/authors (at least specifically in addition to the sampled file data). I'm guessing it'd need to be moved to some common folder outside of the stats plugin for this to use it in a non-spaghetti-code way, though? I'm not sure where that would go. If there's an easy way to read PDF fingerprints than that would be even faster for PDFs that have it (and if any odd PDFs didn't, could fall back to this partial MD5).

By the way, what exactly is the reasoning behind having a sidecar _directory_ rather than just a sidecar file? I know you can have an old version of the file, though I don't understand when those actually gets used. I ask as it would be nice to decrease storage complexity, and it causes other programs like Syncthing to struggle with on Android if you have [lots of folders](https://forum.syncthing.net/t/android-inotify-and-no-root-the-8192-folders-limit/16058/4).

**Comment by ryanwwest** [comment]
I edited my last comment significantly after posting (so initial email notification is outdated). Maybe that's a bad habit.

**Comment by poire-z** [comment]
>  I'm guessing it'd need to be moved to some common folder outside of the stats plugin for this to use it in a non-spaghetti-code way, though? I'm not sure where that would go. 

@hius07 may have that in his todo list, cf: https://github.com/koreader/koreader/pull/10869#issuecomment-1704356923

> If there's an easy way to read PDF fingerprints than that would be even faster for PDFs that have it

Can't help, most of us are not really familiar with MuPDF. But you would indeed need to quickly check how fast it would be to have it "open enough" a PDF to get that fingerprint. It it is slow and needs to read the full book to be able to give you that fingerprint, it may be a showstopper.

> By the way, what exactly is the reasoning behind having a sidecar directory rather than just a sidecar file?

I think it comes from Kindle. Kindle do (or used to do) have/create/use that sidecar directory. The first developpers of KOReader decided to re-use that existing directory (and so, create it on other platforms) to put our own stuff in it.
So, it's a directory, and we can put more stuff in it. We hadn't until recently (we only had metadata.epub.lua and its .old backup), but now we can have custom cover or custom metadata files.

> but I can't grasp the use cases when the old file have something that the new one does not

We don't migrate anything when you switch that location setting (we couldn't know where all the books are). So, on book opening, we try locations in order until we find one - and save to the preferred location on exit.
Also, books may be on a read-only SDCard, so the preferred location may not work, and the other(s) act as a fallback. + the previous older fallback/legacy locations.

We didn't do all that handling for pleasure :) we just tried to not break all the situations we could think about :)

> If backups are a must, then if this were being implemented for the first time I'd find it preferable to keep them all in a separate docsettings/backups/ folder using the same name

The initial choice was not ours - but we tried to handle it, and it has allowed us to do some nice things like fsync() a single directory to get both the .old and .lua sure to be saved sanely to disc. There is quite a history of settings corruption being zero on crash/after reboot. The current situation is quite sane for now - so always risky to go fancy and break what seems to be stable.

> I edited my last comment significantly after posting (so initial email notification is outdated). Maybe that's a bad habit.

You should look at @NiLuJe 's "edited" history on most of his posts :)
(Everytime I tell myself: I'll wait 30mn before reading his comments - and everytime I can't wait and read each "edit" diff as they happen - like I did on your post :)

**Comment by ryanwwest** [comment]
> But you would indeed need to quickly check how fast it would be to have it "open enough" a PDF to get that fingerprint

I learned how PDF.js and Hypothes.is make a fingerprint - tldr it's either from a unique bytestring found in 95% of PDFs or the MD5 of the first 1024 bytes if the bytestring is missing (source: https://www.seanh.cc/2017/11/22/pdf-fingerprinting/). In the latter case, edits such as saving annotations to the file could potentially break the hash depending on where in the file highlight data is saved.

> We hadn't until recently (we only had metadata.epub.lua and its .old backup), but now we can have custom cover or custom metadata files

...And from your other comments, makes sense, I'll have it conform to the current directory requirements then (less work too). 

Though I think the naming of `839136353a953a538b122b297bb2fe92.sdr` is necessary to quickly identify the folder by hash. I guess you could read every sdr folder's contents to pull out a partial MD5 hash from within but that seems slow and unnecessary. I'm toying with the idea of alternatively identifying based on title or author if no hashes are found but this also seems slow and maybe unreliable. Having the hash as the filename decreases readability but this is already inside an internal folder meant for KOReader and not users—maybe you could have `<hash>_<title>.sdr/` and use `<title>` for readability and potential future alternative identification, but then you encroach on the silly Windows 256 filepath character limit.

> and everytime I can't wait and read each "edit" diff as they happen - like I did on your post :)

Lol, sorry. There's too much anticipation/dopamine in pressing that big green 'comment' button and then frantically rereading/rethinking it, and I don't want to annoy people with 10 email notifies.

**Comment by Frenzie** [comment]
If you want to learn more about MuPDF the best way is https://mupdf.com/docs/mupdf-explored.html

Besides that there are the header files.

**Comment by NiLuJe** [comment]
> but then you encroach on the silly Windows 256 filepath character limit.

On a related matter, moving everything to a single directory will also introduce performance issues (on most filesystems, at least) for large libraries. The usual approach for hash based directories like those is a nested layout, with numerical folders at the head (e.g., `0/18/47/<hash>.sdr/`). Kobo's own cache works something like that (the hash part is a hash of the full path, IIRC). So does nginx's (completely random example, I just happened to deal with it recently ;p).

I don't quite recall how the numbers are generated on Kobo, but it's stable, so looking at the Kobo calibre plugin should explain that.

**Comment by ryanwwest** [comment]
Well I tried and failed to get the PDF ID value from mupdf. As I mention above, this is possible by getting the contents of the PDF 'trailer' which is the final xref object in the file, which is possible with the `mutool show <file.pdf>` utility and the needed functions to call from are [here](https://github.com/ArtifexSoftware/mupdf/blob/79465406bd8591ff9a20b0a7e3ee97a3d04fd8b8/source/tools/pdfshow.c#L59) (for printed output) or similar to [here](https://github.com/pymupdf/PyMuPDF/blob/6b7286eb9d6a99125e21b2ad82910a63773c9602/src/extra.i#L1333-L1348C1) (for retrieving the actual value). I got a good way into implementing the latter portion in `koreader-base`'s FFI code in e.g. https://github.com/koreader/koreader-base/blob/master/ffi/mupdf.lua and https://github.com/koreader/koreader-base/blob/master/wrap-mupdf.h, to the point where it compiled and I could call newly wrapped mupdf C functions from [docsettings.lua](https://github.com/koreader/koreader/blob/master/frontend/docsettings.lua). But I got stuck trying to add [FzOutput](https://github.com/ArtifexSoftware/mupdf/blob/79465406bd8591ff9a20b0a7e3ee97a3d04fd8b8/include/mupdf/fitz/output.h#L102-L117C1) into the FFI code as needed [here](https://github.com/pymupdf/PyMuPDF/blob/6b7286eb9d6a99125e21b2ad82910a63773c9602/src/extra.i#L1310) and [here](https://github.com/pymupdf/PyMuPDF/blob/6b7286eb9d6a99125e21b2ad82910a63773c9602/src/extra.i#L1496-L1505) and have run out of time. I'm sure it can be done but don't know enough about how FFI conversion works to do it in my extra time.

Since the PDF ID is nice but not required, when I have more free time I'll instead move to using the partial MD5 as the only identification method for now for all filetypes, which is needed no matter what.

**Comment by NiLuJe** [comment]
Make sure to push that in a branch somewhere, I might drive-by the FFI declarations for you ;).

**Comment by ryanwwest** [comment]
> Make sure to push that in a branch somewhere, I might drive-by the FFI declarations for you ;).

Okay :) Here's the mess so far: mainly https://github.com/ryanwwest/koreader-base/tree/10892-pdf-hash-id, and just a bit of  https://github.com/ryanwwest/koreader/tree/10892-pdf-hash-id.

**Comment by ryanwwest** [comment]
I have a basic working branch for the partialMD5-hashed based SDR storage at https://github.com/ryanwwest/koreader/tree/10892-hashed-metadata. It seems to work with various files across any directory including multiple copies of the same file (that share the same metadata). It also seemed to migrate sdr folders from different places when updating the setting in the UI as per I think [DocSettings.lua:UpdateLocation](https://github.com/ryanwwest/koreader/blob/63785f7e3716b2a35ff225eb3620b765ab0f1e17/frontend/docsettings.lua#L428) Note this is very minimal - no tests are added, I think docsettings.lua would need more changes, better naming and descriptions, etc. 

I noticed that there are some file manager operators for moving and copying files that include moving and copying the metadata sidecar folders. I can't tell if these were already updated to work with both the `dir` and `doc` options when @hius07 did https://github.com/koreader/koreader/pull/10074 or if that's still necessary (for both that and this).

I'm not sure if @NiLuJe 's mention of nested numerical folders like `0/18/47/<hash>.sdr/` would be good to implement here - I'm not opposed to it but don't have a lot of time to figure out Lua folders (I'm fairly new to the language still).

The bigger question is whether anyone is opposed to this or sees issues beyond code quality in it being eventually accepted as a PR. This hash-based metadata storage option has both pros and cons, the biggest con being if files are modified, metadata is lost. I don't think this is worse than any of the current options and would be greatly improved if PDF IDs were eventually added (but then some sort of migration might be necessary if they aren't introduced simultaneously), but even then it isn't fixed. 

The best file-modification solution I can think of is having some file or SQL-based mapping of the each partial MD5 hash to its original embedded document title and/or author (and no, this is not the filename). Then if a hash isn't found, a secondary lookup could be done on that mapping using the document title/author. But this seems bad because that secondary lookup would 99.9% of the time be unnecessary as the user is just looking at a new pdf/epub. An alternative could be store the same embedded title or author info, possibly truncated, in the foldername like `<partialmd5>-<title/author>.sdr/` and then lookup is much easier, but that messes up nested numerical folders (`0/18/47/<hash>.sdr/`), makes folder names long, and in both cases the title/author data would be different for each type of document (maybe alleviated if KOReader can already get us this info). This is a thought exercise to see if there are better solutions (a very far-fetched one is link to Zotero and have it manage the context for us), but maybe it's futile and better to just not use this option if you'll be modifying files.

**Comment by NiLuJe** [comment]
> I'm not sure if @NiLuJe 's mention of nested numerical folders like `0/18/47/<hash>.sdr/` would be good to implement here - I'm not opposed to it but don't have a lot of time to figure out Lua folders (I'm fairly new to the language still).

I'm afraid the performance consequences of not doing it would be fairly dramatic, especially given our target platforms (i.e., FAT 32, terrible storage devices and crappy CPUs).

There's no such thing as a "Lua folder", so I'm not sure what would be the hold up ;).

**Comment by NiLuJe** [comment]
As for the actual question, you probably don't want to hear my answer, because I don't necessarily agree with the need for such an approach.

On the other hand, I also don't necessarily care very much about it, so, as long as the code isn't horribly invasive, I'm fairly neutral about the whole thing ;).

**Comment by hius07** [comment]
(1) We use `DocSettings:hasSidecarFile` to determine whether a document has been opened.
To keep this functionality `DocSettings:getDocSidecarFile` must check the new sdr hash-based location, even when the current setting is `doc` or `dir`.

(2) BUT, taking into account the speed impact, I'd propose to add a new checkbox `[ ] Enable hash-based book metadata location`, default to unchecked. To disable this feature totally everywhere, including `DocSettings:hasSidecarFile`.

(3) md5 is also calculated in
https://github.com/koreader/koreader/blob/55daf9fbf4d315a898f6c3826c82a4b904f77f19/frontend/document/document.lua#L158
So I think `DocSettings:partialMd5(file)` should be moved to `utils` and used everywhere.

**Comment by NiLuJe** [comment]
Right, I was a tad afraid of something like 1., but my tired brain kinda stopped at "eh, might not be a such a problem for previously opened files", but, yeah, that's a problem for *unopened* files, hence something like 2.

Which I'm not necessarily a fan of, but it's pretty much that or nothing, so, err, think carefully ? ;).

**Comment by ryanwwest** [comment]
> (2) BUT, taking into account the speed impact, I'd propose to add a new checkbox [ ] Enable hash-based book metadata location, default to unchecked. To disable this feature totally everywhere, including DocSettings:hasSidecarFile.

I think I understand your point—to avoid always calculating partial hashes, `[ ] Enable hash-based book metadata location` is unchecked until the user specifically decides to enable hashed metadata (and before that, any hashed metadata folders will be ignored). This is slightly separate from just having the option to save to hashed folders by default (from my original message).

Perhaps we could put this setting at `Cog -> Document -> Enable hash-based book metadata location` right below `Book Metadata Location`, and automatically enable it if the user changes to hash-based storage in `Book Metadata Location`, and if the user changes `Book Metadata Location` back to a non-hash storage option, gives the user an info warning that they can uncheck it to increase performance if they've migrated (or let KOReader migrate) their hash-based sdr folders already, and suggest that they use `File Browser -> Cog -> Move Book Metadata` for migration and then uncheck thereafter. Or even have `Move Book Metadata` offer to uncheck if upon completion, 1) the metadata storage option isn't hash-based, 2) there are no remaining hash.sdr folders, and 3) the option is checked.

Also, is there any way to alter the `book metadata location` setting prior to actually opening some document? It's always confused me a bit how there are some KOReader-wide settings, not specific to a document, but that can only be accessed after a document is opened and thus give the impression that they might apply to that particular document.

**Comment by ryanwwest** [comment]
Tbh I wish we could do without the extra checkbox - but I don't know what to do otherwise if the user switches away from hash-based storage and has hashed sdr folders. We could instead force the migration to happen immediately if the user switches away to forego the checkbox... But I'm not sure if the user is just casually playing with options.

**Comment by NiLuJe** [comment]
> to avoid always calculating partial hashes

Technically, the hash part is not what bothers me most (I mean, it *does* bother me, also ;p), actually, it's the aforementioned I/O issues, because to see if a file has been opened, we try to poke at *all* the sidecar locations/formats in order.

Hence what I said above: https://github.com/koreader/koreader/issues/10892#issuecomment-1728153155

**Comment by NiLuJe** [comment]
On the other hand, now that I think about it, the centralized filename based scheme is *also* terrible for I/O, so, eeeh.

**Comment by hius07** [comment]
> Also, is there any way to alter the `book metadata location` setting prior to actually opening some document?

Yes, this menu item is available in the file browser.

**Comment by ryanwwest** [comment]
> I've never really understood the issues people seemed to have with the side-by-side folders
I use Zotero 

Here's my use case, @NiLuJe . I use Zotero to manage all documents, both academic PDFs and EPUBs/HTMLs and whatever else, and Zotero has an option to store the actual files (not its file item metadata) in a folder of your choice which allows for easy syncing with tools like Syncthing (see screenshot). This means I have 1000+ items and many subfolders in this folder and I sync them between 10+ devices and can locally access them on any of them.

![image](https://github.com/koreader/koreader/assets/31378238/124d78fb-6828-48e0-81bc-f525badd1977)

This also means that if I use KOReader's default side-by-side folder for sidecar storage, then on any file browser outside of KOReader's own, I see 100s of long folder names which make file navigation challenging, especially across various different file browser apps and OSes. Granted having 1000 items is also challenging so I have subfolders, but the sdr folders scattered everywhere make things much more difficult. Luckily, I get my KOReader annotations everywhere so I deal with it but it gets harder as the sdr folders build. It's also hard since Zotero likes to rename attachment files sometimes or I want to do some folder organization and have to deal with moving metadata and sometimes forget. So the first two options aren't great for me.

@hius07 It was right under my nose! Thanks for pointing it out.

**Comment by NiLuJe** [comment]
While I've always thought the syncing argument to be perfectly bogus (just teach your syncing stuff to the right thing already); I also never really browse the FS on *any* of my devices (mainly because calibre, but I'm pretty sure I wouldn't really browse 'em either way), so that makes that a non-issue in my case, but, okay, fair enough ;).

**Comment by ryanwwest** [comment]
> While I've always thought the syncing argument to be perfectly bogus (just teach your syncing stuff to the right thing already);

I'd love to hear what you mean by that sometime, if only Github had DMs. Syncthing can be annoying in many ways, but it's the best solution I've found so far given that I love having unnecessarily too many devices.

**Comment by NiLuJe** [comment]
Well nothing fancy, really, just point it at the right stuff, with sane whitelists/blacklists if you need something fancier.

Granted, I'm more of a rsync guy myself, which possibly allows me funkier setups.

**Comment by poire-z** [comment]
(Just had a quick read of the above discussion, might has missed some ideas.)

> The bigger question is whether anyone is opposed to this or sees issues beyond code quality in it being eventually accepted as a PR 

I don't think I've read anything about the probably slow cases I mentionned. Have you tried them ?

Back to the top post, you mentionned that our "doc" option cluters directories, while our "dir" option is nice because it doesn't :) but it has issues like not supporting renaming parent dirs or filename. While our "doc" option supports that, or has the cluttering dir near the filename to remind you that you should rename them too :) So, your 3rd option feels there just because you don't like our .sdr near the book, in spite of all the issues it avoids :)

So, clutter & no issue - vs. no clutter and slowness.
I think the cluttering is little vs. all the problem you may have to solve.

Anyway:
For the checkbox `[ ] Enable hash-based book metadata location`, I think for most people who will not try&need this feature, they will always have that "hash based directory" empty - so, an early check for it empty and never doing the additional work of computing doc checksum would be enough for all users. Those having a few hash-based-sdr in that hash-based-maindir will get it slow, until these last documents are opened and migrated (or these sdr manually deleted).

**Comment by hius07** [comment]
> they will always have that "hash based directory" empty

It is the same koreader/docsettings
https://github.com/ryanwwest/koreader/blob/63785f7e3716b2a35ff225eb3620b765ab0f1e17/frontend/docsettings.lua#L79

**Comment by ryanwwest** [comment]
> I don't think I've read anything about the probably slow cases I mentionned. Have you tried them ?

Not yet - I'm planning to look at this more after implementing more of the needed changes in `docsettings.lua` which someone mentioned somewhat higher up.

> I think the cluttering is little vs. all the problem you may have to solve.

There are tradeoffs to each. Hundreds of extra `.sdr` folders everywhere is becoming too much of one for me hence this FR. I think there's some better solution than both of them but don't know what it is.

> It is the same koreader/docsettings

This could be different, doesn't matter to me. Then maybe @poire-z's suggestion works—so are you suggesting that we forego the checkbox and it just depends on if there are any files in the hash-dir?

**Comment by ryanwwest** [comment]
I [made some changes](https://github.com/ryanwwest/koreader/tree/10892-hashed-metadata) updating `getDocSidecarFile` to get the hashed sidecar directory. Also, now the hash is only calculated if a user has at least one existing hashed sidecar directory (which are now stored in a separate folder). A hash is indeed calculated on each file on the visible page in the file browser this way to display things like reading progress to the side, but it looks like it only calculates them for that page so 10kb read per file * 5-20 files = 50-200kb per page read (maybe opening/closing the files is more performance intensive at this level).

I'm going to figure out how to compile this to APK and put it on Android and test it out a bit.

By the way, are there any instructions to debugging this in VS Code? Spamming `logger.warn()` to get variable values leaves a lot to be desired, even without needing to recompile each time...

**Comment by hius07** [comment]
`not util.isEmptyDir(DOCSETTINGS_HASH_DIR)`
Not needed, impossible situation.

`if lfs.attributes(DOCSETTINGS_HASH_DIR, "mode") == "directory"`
I'd move this check to `DocSettings:getDocSidecarFile` to avoid extra `lfs` call with "".
And put all this check of hash-based files at the end, as the least frequent option.

**Comment by Frenzie** [comment]
> By the way, are there any instructions to debugging this in VS Code?

You can do something like this with ZeroBrane. Whether that works with the MobDebug adapter for VS Code is left as an exercise to people who are extremely attached to VS Code.

```diff
diff --git a/reader.lua b/reader.lua
index 54d1793b..110bccfe 100755
--- a/reader.lua
+++ b/reader.lua
@@ -255,6 +255,10 @@ local function retryLastFile()
     }
 end
 
+package.path = package.path .. ";/opt/zbstudio/lualibs/mobdebug/?.lua"
+require('mobdebug').start()
+
+
 -- Start app
 local exit_code
 if file then
```

**Comment by ryanwwest** [comment]
> not util.isEmptyDir(DOCSETTINGS_HASH_DIR)
> Not needed, impossible situation.

@hius07 Could you elaborate on 'impossible'? Do you mean this is already checked with `lfs.attributes(DOCSETTINGS_HASH_DIR, "mode") == "directory"` or that such a check is unnecessary? I was considering if a user started using KOReader on the `doc` setting, eventually switched to the `hash` setting, decided they didn't like it and moved back to `doc`, but at that point had a couple straggler hash sidecar files that still need checking, so need to check if the hash directory exists AND there are files inside (unless you're suggesting that the directory's existence alone is enough to warrant partial hash calculation). 


> I'd move this check to DocSettings:getDocSidecarFile to avoid extra lfs call with "".

There are two callers of `maybeGetSidecarHashDirAndFilepath` so I don't think I can move the directory existence check out—but I could potentially put https://github.com/ryanwwest/koreader/blob/e00e5c2fa65be5aa6b1bb14cc16b6e13aa1bdfa3/frontend/docsettings.lua#L115 _inside_ `maybeGetSidecarHashDirAndFilepath` so it is only run if the directory was there to lower the normal `lfs` call count. But I might be misunderstanding your point. I did move this logic to the bottom of `getDocSidecarFile`.

@Frenzie What editor/debugger do you use?

**Comment by hius07** [comment]
> Could you elaborate on 'impossible'?

https://github.com/koreader/koreader/blob/cafe565adeb9a5c2d534c6c467babadefd1dcaeb/frontend/docsettings.lua#L346
When docsettings folder becomes empty, it is removed.
If docsettings folder exists, it is not empty.

**Comment by hius07** [comment]
I propose to check existence of DOCSETTINGS_HASH_DIR right in `DocSettings:getDocSidecarFile()` (and do nothing if it is absent) because it is the most frequently used method.

**Comment by Frenzie** [comment]
> @Frenzie What editor/debugger do you use?

It varies, mainly Geany and Kate. But see #10708 for discussion.

**Comment by ryanwwest** [comment]
@hius07 I implemented your suggestions in the latest commit—I don't think a separate checkbox is needed now. Also, @Frenzie's suggestion of nested hash subdirectories is added to mitigate too many files in one directory.

I haven't gotten an Android build yet to test on ereaders from errors, but the emulator on MacOS doesn't seem to be slower than regular KOReader on MacOS (in file browser only 5-9 files need hashes calculated per page which is still quick).

Should 'book metadata location' be renamed to something more general like 'book metadata storage method'? The 'about' info popup disambiguates how the method changes, so maybe it can stay the same but not sure.

![image](https://github.com/koreader/koreader/assets/31378238/0f1d6dba-7bd1-493f-b8fc-243a2c94e969)

**Comment by hius07** [comment]
Agree, the checkbox is not needed.

I think it is good in general. Please add removing empty hash_sidecar_dir to `purge()` and `removeSidecarDir()`.
There are some minor comments that would be convenient to discuss having the code in the main repository PR.

**Comment by Frenzie** [comment]
> the emulator on MacOS doesn't seem to be slower than regular KOReader on MacOS

That's fairly meaningless unless it was in Valgrind or something. :-) That's all several orders of magnitude faster than some poor ereader.

But at this point I've lost track regarding when there's extra I/O and processing.

**Comment by ryanwwest** [comment]
I'll put this into a PR. Maybe that will autogenerate the Android build for me to test since I can't figure it out lol. Before the PR I'll also make partialMD5 common for other functionality too.

Is `common_settings_menu_table.lua` is there a trick to getting `UIManager.show(InfoMessage:new{text = "some text"},)` to actually render an info dialog? I'm trying to add a warning message if enabling either the 'Save Document (write highlights into PDF)' or hash-based metadata while the other is already on so users know (and maybe will do the same for calibre book imports) but I keep getting error ["attempted to show a nil widget](https://github.com/ryanwwest/koreader/blob/d1bbdb5dad836c3f838b75215f0fae1f81d3a6a6/frontend/ui/uimanager.lua#L137) when doing so even though the line above looks similar to other working ones...

And also about this, we could alternatively just make these settings mutually exclusive entirely. I don't know if settings can be 'greyed out' but maybe just disallow users from selecting both at once?

**Comment by hius07** [comment]
Have you seen this
https://github.com/koreader/koreader/blob/741302445b5e523d71e84795a16dd8a3fe6e7d45/frontend/document/document.lua#L146-L193

**Comment by ryanwwest** [comment]
Ah that's basically a copy of the partialMD5. I can use that instead.

**Comment by NiLuJe** [comment]
That was possibly an attempt at limiting the eventuality of a hash collision ;).

**Comment by ryanwwest** [comment]
PR ^

**Comment by poire-z** [comment]
> A hash is indeed calculated on each file on the visible page in the file browser this way to display things like reading progress to the side, but it looks like it only calculates them for that page so 10kb read per file * 5-20 files = 50-200kb per page read (maybe opening/closing the files is more performance intensive at this level). This doesn't seem like an issue so far, but needs more testing.

I mentionned this case you may not have checked:

> also, some filebrowser sort options (and probably other things) need to know the above for all books in a directory before it can order them and show the first page - so the need to compute the md5 of all the book :/

These options may be the `Sort by: percent - unopened first/last`, which I think would need to open ALL the docsettings of the file in the directory at start, and not only 5-20 files per page.

**Comment by ryanwwest** [comment]
> These options may be the Sort by: percent - unopened first/last, which I think would need to open ALL the docsettings of the file in the directory at start, and not only 5-20 files per page.

Thanks, I missed this. Hmm, problematic for my use case with a folder with many hundreds of books. A per-device solution after the first one is a persistent cache of mappings from path to hash like you mentioned at https://github.com/koreader/koreader/pull/10945#discussion_r1340456884 (but persists across app restarts), but that's still not really ideal. 

> And may be also the options in Classic mode settings > Show opened/new files in bold.

I'm not familiar with this.

**Comment by ryanwwest** [comment]
I tested the "Sort by: percent - unopened first/last" file browser mode on my Boox Leaf 2 which is known as a rather underpowered ereader, and had it sort a directory with ~250 items. There wasn't any discernible difference in loading or browsing times compared to other modes. It was a complete nonissue - maybe with a few thousand items it would become a problem?

**Comment by poire-z** [comment]
Well, I have no idea - it's for you to try to gather stats of how long a partialMD5 computations takes (we have some time.since & co methods , cf frontend/apps/reader/readerui.lua, to gather times, so sum / count can give you an estimate - to do after a fresh reboot - and the same after a koreader restart while things still cached by the OS).
If you get each < 10ms on most devices, it might be ok (2 second waiting for a 200 books directory might be ok?)

**Comment by ryanwwest** [comment]
I used https://www.loc.gov/item/2020445568/ which has 1000 varying-length PDFs to further test this on a few devices, the slowest being the Boox Leaf 2. The time taken to open files is virtually unchanged no matter the document I've tested since a single file's 10kb hash calculation stays the same, and I feel like we could probably remove the warning for this near-zero additional load time.

The file browser "Sort by: percent - unopened first/last" mode is different (though I think it's still fine). With either hash-based SDR storage (and scanning) enabled or disabled, after a restart it took my slowest device ~6.1 seconds over a few runs to load this 1000-file directory and roughly same time to load it when sorting by e.g. date modified (file-based attributes), perhaps .1 seconds faster.

When enabling hash-based SDR storage and sorting by date modified the time remained the same. However, when "Sort by: percent - unopened first/last" was enabled it increased to about ~7.9 seconds. So, there is definitely a difference for this slower device, but it seems reasonable to to warn the user and keep it as is, given that the user already experiences a significant delay if they are dealing with directories this filled (which I think should be expected once you start going above ~500 anyway) and since the additional delay is relatively not that much.
